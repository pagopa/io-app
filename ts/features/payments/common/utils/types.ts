export type AlertVariant = "error" | "success" | "warning" | "info";

export type PaymentsBackoffRetryValue = {
  retryCount: number;
  allowedRetryTimestamp?: number;
};
