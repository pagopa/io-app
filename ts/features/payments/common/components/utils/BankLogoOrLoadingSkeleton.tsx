import { IOColors, IOSkeleton } from "@pagopa/io-app-design-system";
import { useEffect, useState } from "react";
import { Image } from "react-native";
import { getBankLogosCdnUri } from "../../../../../components/ui/utils/strings";

type BankLogoOrSkeletonProps = {
  abiCode?: string;
  dimensions: { height: number; width: number };
  placeHolderColor?: IOColors;
  imageA11yLabel?: string;
};

/**
 *component that renders a bank logo based on the abiCode,
 * and displays a loader while the image is loading
 * @param abiCode
 * @param dimensions - height and width of the image and placeholder
 * @param placeHolderColor - color of the placeholder
 */
export const BankLogoOrSkeleton = ({
  abiCode,
  dimensions,
  placeHolderColor = "grey-200",
  imageA11yLabel
}: BankLogoOrSkeletonProps) => {
  const [imageUrl, setImageUrl] = useState<string | undefined>(undefined);
  const { height, width: maxWidth } = dimensions;
  const [width, setWidth] = useState<number>(maxWidth);
  useEffect(() => {
    // we pre-fetch the image to avoid having to render both items
    // at the same time, which looks like an untidy hack
    if (abiCode === undefined) {
      return;
    }
    fetchBlob(getBankLogosCdnUri(abiCode))
      .then(blob => {
        const url = URL.createObjectURL(blob);
        // to make sure the ratio is correct, we
        // calculate it ourselves, else we risk misalignments
        Image.getSize(url, (width, imgHeight) => {
          const ratio = width / imgHeight;
          setWidth(ratio * height);
        });
        setImageUrl(url);
      })
      .catch(_ => null);
  }, [abiCode, height]);

  return imageUrl !== undefined ? (
    <Image
      accessibilityIgnoresInvertColors
      accessible
      accessibilityLabel={imageA11yLabel}
      source={{ uri: imageUrl }}
      style={{
        height,
        width,
        resizeMode: "contain"
      }}
    />
  ) : (
    <IOSkeleton
      color={IOColors[placeHolderColor]}
      shape="rectangle"
      radius={8}
      height={height}
      width={width}
    />
  );
};
async function fetchBlob(url: string) {
  const response = await fetch(url);
  if (response.status === 200) {
    return response.blob();
  } else {
    throw new Error();
    // error is caught and irrelevant
  }
}
