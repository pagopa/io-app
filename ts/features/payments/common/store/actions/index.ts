import { PaymentsMethodDetailsActions } from "../../../details/store/actions";
import { PaymentsHistoryActions } from "../../../history/store/actions";
import { PaymentsOnboardingActions } from "../../../onboarding/store/actions";
import { PaymentsCheckoutActions } from "../../../checkout/store/actions";
import { PaymentsTransactionActions } from "../../../transaction/store/actions";
import { PaymentsWalletActions } from "../../../wallet/store/actions";

export type PaymentsActions =
  | PaymentsOnboardingActions
  | PaymentsMethodDetailsActions
  | PaymentsCheckoutActions
  | PaymentsTransactionActions
  | PaymentsHistoryActions
  | PaymentsWalletActions;
