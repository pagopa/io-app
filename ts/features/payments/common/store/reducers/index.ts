import { combineReducers } from "redux";
import { PersistPartial } from "redux-persist";
import detailsReducer, {
  PaymentsMethodDetailsState
} from "../../../details/store/reducers";
import historyReducer, {
  PaymentsHistoryState
} from "../../../history/store/reducers";
import onboardingReducer, {
  PaymentsOnboardingState
} from "../../../onboarding/store/reducers";
import paymentReducer, {
  PaymentsCheckoutState
} from "../../../checkout/store/reducers";
import transactionReducer, {
  PaymentsTransactionState
} from "../../../transaction/store/reducers";

export type PaymentsState = {
  onboarding: PaymentsOnboardingState;
  details: PaymentsMethodDetailsState;
  payment: PaymentsCheckoutState;
  transaction: PaymentsTransactionState;
  history: PaymentsHistoryState & PersistPartial;
};

const paymentsReducer = combineReducers({
  onboarding: onboardingReducer,
  details: detailsReducer,
  payment: paymentReducer,
  transaction: transactionReducer,
  history: historyReducer
});

export default paymentsReducer;
