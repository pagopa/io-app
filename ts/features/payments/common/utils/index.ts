import {
  IOLogoPaymentType,
  IOPaymentLogos,
  ListItemTransactionStatusWithBadge
} from "@pagopa/io-app-design-system";
import * as O from "fp-ts/lib/Option";
import { pipe } from "fp-ts/lib/function";
import I18n from "i18n-js";
import _ from "lodash";
import { Bundle } from "../../../../../definitions/pagopa/ecommerce/Bundle";
import { WalletApplicationStatusEnum } from "../../../../../definitions/pagopa/walletv3/WalletApplicationStatus";
import { WalletInfo } from "../../../../../definitions/pagopa/walletv3/WalletInfo";
import { PaymentSupportStatus } from "../../../../types/paymentMethodCapabilities";
import { getDateFromExpiryDate, isExpiredDate } from "../../../../utils/dates";
import { WalletPaymentPspSortType } from "../../checkout/types";
import { PaymentCardProps } from "../components/PaymentCard";
import { UIWalletInfoDetails } from "../types/UIWalletInfoDetails";
import { findFirstCaseInsensitive } from "../../../../utils/object";
import { WalletCard } from "../../../newWallet/types";

/**
 * A simple function to get the corresponding translated badge text,
 * based on the transaction status.
 */

export const getBadgeTextByTransactionStatus = (
  transactionStatus: ListItemTransactionStatusWithBadge
) => {
  switch (transactionStatus) {
    case "failure":
      return I18n.t("global.badges.failed");
    case "cancelled":
      return I18n.t("global.badges.cancelled");
    case "reversal":
      return I18n.t("global.badges.reversal");
    case "pending":
      return I18n.t("global.badges.onGoing");
    default:
      return "";
  }
};

/**
 * Check if the given payment method is expired
 * right(true) if it is expired, right(false) if it is still valid
 * left if expiring date can't be evaluated
 * @param paymentMethod
 */
export const isPaymentMethodExpired = (
  details?: UIWalletInfoDetails
): boolean =>
  pipe(
    details?.expiryDate,
    O.fromNullable,
    O.chainNullableK(getDateFromExpiryDate),
    O.chainNullableK(isExpiredDate),
    O.getOrElse(() => false)
  );

/**
 * true if the given paymentMethod supports the given walletFunction
 * @param paymentMethod
 * @param walletFunction
 */
export const hasApplicationEnabled = (
  paymentMethod: WalletInfo | undefined,
  walletApplication: string
): boolean =>
  paymentMethod !== undefined &&
  paymentMethod.applications.some(
    application =>
      application.name === walletApplication &&
      application.status === WalletApplicationStatusEnum.ENABLED
  );
/**
 * return true if the payment method has the payment feature
 */
export const hasPaymentFeature = (paymentMethod: WalletInfo): boolean =>
  hasApplicationEnabled(paymentMethod, "PAGOPA");

/**
 * Check if a payment method is supported or not
 * If the payment method have the enableable function pagoPA, can always pay ("available")
 * "available" -> can pay
 * "arriving" -> will pay
 * "notAvailable" -> can't pay
 * "onboardableNotImplemented" -> can onboard a card that can pay but is not yet implemented
 */
export const isPaymentSupported = (
  paymentMethod: WalletInfo
): PaymentSupportStatus => {
  const paymentSupported: O.Option<PaymentSupportStatus> = hasPaymentFeature(
    paymentMethod
  )
    ? O.some("available")
    : O.none;

  const notAvailableCustomRepresentation = O.some(
    "notAvailable" as PaymentSupportStatus
  );

  return pipe(
    paymentSupported,
    O.alt(() => notAvailableCustomRepresentation),
    O.getOrElseW(() => "notAvailable" as const)
  );
};

export const WALLET_PAYMENT_TERMS_AND_CONDITIONS_URL =
  "https://www.pagopa.gov.it/it/prestatori-servizi-di-pagamento/elenco-PSP-attivi/";

/**
 * Function that returns a sorted list of psp based on the given sortType
 * The sortType can be: "name", "amount" or "default"
 */
export const getSortedPspList = (
  pspList: ReadonlyArray<Bundle>,
  sortType: WalletPaymentPspSortType
) => {
  switch (sortType) {
    case "name":
      return _.orderBy(pspList, psp => psp.bundleName);
    case "amount":
      return _.orderBy(pspList, psp => psp.taxPayerFee);
    case "default":
    default:
      return _.orderBy(pspList, ["onUs", "taxPayerFee"]);
  }
};

export const getPaymentCardPropsFromWalletInfo = (
  wallet: WalletInfo
): PaymentCardProps => {
  const details = wallet.details as UIWalletInfoDetails;
  const isExpired = isPaymentMethodExpired(details);

  return {
    hpan: details.lastFourDigits,
    brand: details.brand,
    expireDate: getDateFromExpiryDate(details.expiryDate),
    holderEmail: details.maskedEmail,
    holderPhone: details.maskedNumber,
    isExpired
  };
};

export const getPaymentLogoFromWalletDetails = (
  details: UIWalletInfoDetails
): IOLogoPaymentType | undefined => {
  if (details.maskedEmail !== undefined) {
    return "payPal";
  } else if (details.maskedNumber !== undefined) {
    return "bancomatPay";
  } else {
    return pipe(
      details.brand,
      O.fromNullable,
      O.chain(findFirstCaseInsensitive(IOPaymentLogos)),
      O.map(([logoName, _]) => logoName as IOLogoPaymentType),
      O.toUndefined
    );
  }
};

export const mapWalletIdToCardKey = (walletId: string) => `method_${walletId}`;

export const mapWalletsToCards = (
  wallets: ReadonlyArray<WalletInfo>
): ReadonlyArray<WalletCard> =>
  wallets.map<WalletCard>(wallet => ({
    ...getPaymentCardPropsFromWalletInfo(wallet),
    key: mapWalletIdToCardKey(wallet.walletId),
    type: "payment",
    category: "payment",
    walletId: wallet.walletId
  }));

/**
 * Function that returns a formatted payment notice number
 * by placing two spaces between every four numbers
 */
export const formatPaymentNoticeNumber = (noticeNumber: string) =>
  noticeNumber.replace(/(\d{4})/g, "$1  ").trim();
