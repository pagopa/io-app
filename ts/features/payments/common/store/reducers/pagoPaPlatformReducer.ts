import * as pot from "@pagopa/ts-commons/lib/pot";
import { getType } from "typesafe-actions";
import { Action } from "../../../../../store/actions/types";
import { NetworkError } from "../../../../../utils/errors";
import {
  clearPaymentsPendingActions,
  paymentsGetPagoPaPlatformSessionTokenAction,
  paymentsResetPagoPaPlatformSessionTokenAction,
  savePaymentsPendingAction
} from "../actions";

export type PaymentsPagoPaPlatformState = {
  sessionToken: pot.Pot<string, NetworkError>;
  pendingActions: Array<Action>;
};

const INITIAL_STATE: PaymentsPagoPaPlatformState = {
  sessionToken: pot.none,
  pendingActions: []
};

const reducer = (
  state: PaymentsPagoPaPlatformState = INITIAL_STATE,
  action: Action
): PaymentsPagoPaPlatformState => {
  switch (action.type) {
    case getType(paymentsGetPagoPaPlatformSessionTokenAction.request):
      return {
        ...state,
        sessionToken: pot.toLoading(state.sessionToken)
      };
    case getType(paymentsGetPagoPaPlatformSessionTokenAction.success):
      return {
        ...state,
        sessionToken: pot.some(action.payload.token as string)
      };
    case getType(paymentsGetPagoPaPlatformSessionTokenAction.failure):
      return {
        ...state,
        sessionToken: pot.toError(state.sessionToken, action.payload)
      };
    case getType(paymentsResetPagoPaPlatformSessionTokenAction):
      return {
        ...state,
        sessionToken: pot.none
      };
    // ADD PENDING ACTION WHILE REFRESHING THE PAGOPA TOKEN
    case getType(savePaymentsPendingAction):
      const actionToSave = action as ReturnType<
        typeof savePaymentsPendingAction
      >;
      return {
        ...state,
        pendingActions: [
          ...state.pendingActions,
          actionToSave.payload.pendingAction
        ]
      };
    case getType(clearPaymentsPendingActions):
      return {
        ...state,
        pendingActions: []
      };
  }
  return state;
};

export default reducer;
