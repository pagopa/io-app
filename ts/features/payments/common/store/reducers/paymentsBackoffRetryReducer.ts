import { getType } from "typesafe-actions";
import { Action } from "../../../../../store/actions/types";
import {
  clearPaymentsBackoffRetry,
  increasePaymentsBackoffRetry
} from "../actions";
import {
  PAYMENTS_BACKOFF_SECONDS_DELAYS,
  SECONDS_TO_MILLISECONDS
} from "../../utils/backoffRetry";
import { PaymentsBackoffRetryValue } from "../../utils/types";

export type PaymentsBackoffRetryState = {
  [key: string]: PaymentsBackoffRetryValue | undefined;
};

const INITIAL_STATE: PaymentsBackoffRetryState = {};

const reducer = (
  state: PaymentsBackoffRetryState = INITIAL_STATE,
  action: Action
): PaymentsBackoffRetryState => {
  switch (action.type) {
    case getType(increasePaymentsBackoffRetry):
      const currentRetryCount = state[action.payload]?.retryCount ?? 0;
      const retryCount = currentRetryCount + 1;
      const allowedRetryTimestamp =
        Date.now() +
        PAYMENTS_BACKOFF_SECONDS_DELAYS[retryCount] * SECONDS_TO_MILLISECONDS;

      return {
        ...state,
        [action.payload]: {
          retryCount,
          allowedRetryTimestamp
        }
      };
    case getType(clearPaymentsBackoffRetry):
      if (!state[action.payload]) {
        return state;
      }
      return {
        ...state,
        [action.payload]: undefined
      };
  }
  return state;
};

export default reducer;
