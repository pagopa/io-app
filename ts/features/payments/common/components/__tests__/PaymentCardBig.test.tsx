import { default as React } from "react";
import { createStore } from "redux";
import ROUTES from "../../../../../navigation/routes";
import { applicationChangeState } from "../../../../../store/actions/application";
import { appReducer } from "../../../../../store/reducers";
import { GlobalState } from "../../../../../store/reducers/types";
import { renderScreenWithNavigationStoreContext } from "../../../../../utils/testWrapper";
import { PaymentCardBig, PaymentCardBigProps } from "../PaymentCardBig";

describe("PaymentCardBig component", () => {
  const testID = "PaymentCardBigTestID";
  jest.useFakeTimers();
  it(`matches snapshot for loading`, () => {
    const component = renderCardBig({ isLoading: true });
    expect(component).toMatchSnapshot();
  });
  it(`matches snapshot for paypal`, () => {
    const component = renderCardBig({
      cardType: "PAYPAL",
      holderEmail: "someEmail@test.com"
    });
    expect(component).toMatchSnapshot();
  });

  it(`should render a phone number in case of BancomatPay`, () => {
    const component = renderCardBig({
      cardType: "BANCOMATPAY",
      phoneNumber: "1234567890",
      holderName: "holderName",
      testID
    });
    expect(component).not.toBeNull();
    expect(component.queryByText("1234567890")).not.toBeNull();
  });
  it(`should render a skeleton when loading`, () => {
    const component = renderCardBig({
      isLoading: true,
      testID
    });
    expect(component).not.toBeNull();
    expect(component.queryByTestId(`${testID}-skeleton`)).not.toBeNull();
  });
});

function renderCardBig(props: PaymentCardBigProps) {
  const globalState = appReducer(undefined, applicationChangeState("active"));
  return renderScreenWithNavigationStoreContext<GlobalState>(
    () => <PaymentCardBig {...props} />,
    ROUTES.WALLET_HOME,
    {},
    createStore(appReducer, globalState as any)
  );
}
