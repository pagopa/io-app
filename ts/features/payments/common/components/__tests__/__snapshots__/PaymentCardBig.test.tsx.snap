// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentCardBig component matches snapshot for loading 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    Object {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  WALLET_HOME
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSModalScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          gestureResponseDistance={
            Object {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          onGestureCancel={[Function]}
          pointerEvents="box-none"
          sheetAllowedDetents="large"
          sheetCornerRadius={-1}
          sheetExpandsWhenScrolledToEdge={true}
          sheetGrabberVisible={false}
          sheetLargestUndimmedDetent="all"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={1}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "#FFFFFF",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "alignContent": "space-between",
                            "alignItems": "flex-start",
                            "backgroundColor": "#E8EBF1",
                            "borderCurve": "continuous",
                            "borderRadius": 16,
                            "flex": 1,
                            "height": 207,
                            "justifyContent": "space-between",
                            "padding": 24,
                            "width": "100%",
                          }
                        }
                        testID="undefined-skeleton"
                      >
                        <View>
                          <View
                            collapsable={false}
                            style={
                              Object {
                                "opacity": 0.5,
                              }
                            }
                          >
                            <View
                              color="#D2D6E3"
                              height={24}
                              radius={8}
                              style={
                                Object {
                                  "backgroundColor": "#D2D6E3",
                                  "borderRadius": 8,
                                  "height": 24,
                                  "width": 164,
                                }
                              }
                              width={164}
                            />
                          </View>
                          <View
                            style={
                              Object {
                                "height": 16,
                              }
                            }
                          />
                          <View
                            collapsable={false}
                            style={
                              Object {
                                "opacity": 0.5,
                              }
                            }
                          >
                            <View
                              color="#D2D6E3"
                              height={16}
                              radius={8}
                              style={
                                Object {
                                  "backgroundColor": "#D2D6E3",
                                  "borderRadius": 8,
                                  "height": 16,
                                  "width": 117,
                                }
                              }
                              width={117}
                            />
                          </View>
                        </View>
                        <View>
                          <View
                            collapsable={false}
                            style={
                              Object {
                                "opacity": 0.5,
                              }
                            }
                          >
                            <View
                              color="#D2D6E3"
                              height={16}
                              radius={8}
                              style={
                                Object {
                                  "backgroundColor": "#D2D6E3",
                                  "borderRadius": 8,
                                  "height": 16,
                                  "width": 97,
                                }
                              }
                              width={97}
                            />
                          </View>
                          <View
                            style={
                              Object {
                                "height": 4,
                              }
                            }
                          />
                          <View
                            collapsable={false}
                            style={
                              Object {
                                "opacity": 0.5,
                              }
                            }
                          >
                            <View
                              color="#D2D6E3"
                              height={24}
                              radius={8}
                              style={
                                Object {
                                  "backgroundColor": "#D2D6E3",
                                  "borderRadius": 8,
                                  "height": 24,
                                  "width": 164,
                                }
                              }
                              width={164}
                            />
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSModalScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;

exports[`PaymentCardBig component matches snapshot for paypal 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    Object {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  WALLET_HOME
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSModalScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          gestureResponseDistance={
            Object {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          onGestureCancel={[Function]}
          pointerEvents="box-none"
          sheetAllowedDetents="large"
          sheetCornerRadius={-1}
          sheetExpandsWhenScrolledToEdge={true}
          sheetGrabberVisible={false}
          sheetLargestUndimmedDetent="all"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={2}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "#FFFFFF",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "alignContent": "space-between",
                            "alignItems": "flex-start",
                            "backgroundColor": "#E8EBF1",
                            "borderCurve": "continuous",
                            "borderRadius": 16,
                            "flex": 1,
                            "height": 207,
                            "justifyContent": "space-between",
                            "padding": 24,
                            "width": "100%",
                          }
                        }
                      >
                        <Image
                          accessibilityLabel="PayPal"
                          accessible={true}
                          resizeMode="contain"
                          source={
                            Object {
                              "testUri": "../../../img/wallet/payment-methods/paypal-logo.png",
                            }
                          }
                          style={
                            Object {
                              "height": 32,
                              "width": 114,
                            }
                          }
                        />
                        <View
                          style={
                            Object {
                              "height": 8,
                            }
                          }
                        />
                        <Text
                          accessibilityLabel="account: someEmail@test.com"
                          accessible={true}
                          allowFontScaling={false}
                          color="black"
                          defaultColor="black"
                          defaultWeight="SemiBold"
                          ellipsizeMode="tail"
                          font="TitilliumWeb"
                          fontStyle={
                            Object {
                              "fontSize": 18,
                              "lineHeight": 25,
                            }
                          }
                          numberOfLines={1}
                          style={
                            Array [
                              Object {
                                "width": "75%",
                              },
                              Object {
                                "fontSize": 18,
                                "lineHeight": 25,
                              },
                              Object {
                                "color": "#0E0F13",
                                "fontFamily": "Titillium Web",
                                "fontStyle": "normal",
                                "fontWeight": "600",
                              },
                            ]
                          }
                          weight="SemiBold"
                        >
                          someEmail@test.com
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSModalScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
